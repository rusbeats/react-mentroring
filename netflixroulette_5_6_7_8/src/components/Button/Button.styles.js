import constants from '../../utils/constants';
import styled from 'styled-components';

const { colors, getPaddingBySize } = constants;

export const StyledButton = styled.button.attrs(({ size, active, margin, disabled }) => ({
    padding: getPaddingBySize(size),
    backgroundColor: active ? colors.lightPink : colors.grey,
    margin: margin ? '20px' : '0',
    disabled
}))`
    color: ${colors.white};
    background-color: ${props => props.backgroundColor};
    border: none;
    padding: ${props => props.padding};
    text-align: center;
    cursor: pointer;
    text-transform: uppercase;
    font-size: 16px;
    font-weight: bold;
    margin-right: ${props => props.margin};
    border-radius: 3px;
`;
