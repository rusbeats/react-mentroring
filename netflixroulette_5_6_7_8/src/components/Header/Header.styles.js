import constants from '../../utils/constants';
import styled from 'styled-components';

const { colors } = constants;

export const MainHeader = styled.div`
    background-image: linear-gradient(rgba(0, 0, 0, 0.5), rgba(0, 0, 0, 0.5)), url('/static/header-background.png');
    background-repeat: no-repeat;
    background-size: cover;
    height: 300px;
    padding: 20px 125px;
`;

export const HeaderText = styled.div`
    color: ${colors.lightPink};
    font-weight: bold;
    margin-bottom: 50px;
`;

export const SearchHeader = styled.div`
    text-transform: uppercase;
    color: ${colors.white};
    font-weight: bold;
    font-size: 20px;
    margin-bottom: 30px;
`;

export const SearchInput = styled.input.attrs({
    type: 'text',
    placeholder: 'e.g Transformers'
})`
    width: 100%;
    text-indent: 30px;
    padding: 20px 0;
    font-size: 20px;
    background-color: ${colors.black};
    color: ${colors.lightGrey};
    &::-webkit-input-placeholder {
        color: ${colors.lightGrey};
    }
    border: none;
    border-bottom: solid ${colors.lightPink} 2px;
    margin-bottom: 30px;
`;

export const ButtonsWrapper = styled.div`
    display: flex;
    justify-content: space-between;
`;

export const SearchByText = styled.div`
    text-transform: uppercase;
    font-weight: bold;
    color: ${colors.white};
    letter-spacing: 1px;
    margin-right: 20px;
`;

export const SearchByButtons = styled.div`
    display: flex;
    align-items: center;
`;

export const SubHeader = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 60px;
    padding-left: 125px;
    background-color: ${colors.lightGrey};
    font-weight: bold;
    font-size: 20px;
`;

export const Sorting = styled.div`
    display: flex;
    justify-content: space-around;
    width: 40%;
    margin-right: 40px;
`;
