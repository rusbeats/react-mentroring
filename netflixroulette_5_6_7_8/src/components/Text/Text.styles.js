import styled from 'styled-components';

export const StyledText = styled.div`
    color: ${props => props.color};
    cursor: ${props => props.cursor};
    pointer-events: ${props => props.pointerEvents};
`;
