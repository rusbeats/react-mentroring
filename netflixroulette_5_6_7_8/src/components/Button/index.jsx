import React from 'react';
import PropTypes from 'prop-types';
import { StyledButton } from './Button.styles';

const Button = ({ size, active, margin, children, disabled, onClick }) => {
    return (
        <StyledButton size={size} active={active} margin={margin} disabled={disabled} onClick={onClick}>
            {children}
        </StyledButton>
    );
};

StyledButton.displayName = 'Button';

Button.propTypes = {
    size: PropTypes.string,
    active: PropTypes.bool,
    margin: PropTypes.bool,
    children: PropTypes.node,
    disabled: PropTypes.bool,
    onClick: PropTypes.func
};

export default Button;
