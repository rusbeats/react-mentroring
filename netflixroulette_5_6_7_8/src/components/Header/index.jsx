import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import Button from '../Button';
import Text from '../Text';
import { connect } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { toggleSorting, toggleSearchBy } from '../../store/actions';
import { getFilms } from '../../store/api';
import { getSearchUrl } from '../../utils/helpers';
import Link from 'next/link';
import { getHeaderState } from '../../selectors';
import {
    MainHeader,
    HeaderText,
    SearchHeader,
    SearchInput,
    ButtonsWrapper,
    SearchByButtons,
    SearchByText,
    SubHeader,
    Sorting
} from './Header.styles';

export const Header = ({ filmsCount, sortBy, searchBy, toggleSorting, toggleSearchBy, fetchFilms }) => {
    const { t } = useTranslation();
    const sortedByRating = sortBy === 'rating';
    const sortedByReleaseDate = sortBy === 'releaseDate';
    const searchByTitle = searchBy === 'title';
    const searchByGenre = searchBy === 'genre';
    const [searchValue, setSearchValue] = useState('');
    const searchUrl = getSearchUrl(searchValue, searchBy);
    const handleSearchInput = useCallback(e => setSearchValue(e.target.value), [setSearchValue]);
    const handleSearchBy = useCallback(searchBy => () => toggleSearchBy(searchBy), [toggleSearchBy, searchBy]);
    const handleSortBy = useCallback(sortBy => () => toggleSorting(sortBy), [toggleSorting, sortBy]);
    const handleEnter = useCallback(
        e => {
            if (e.key === 'Enter') {
                fetchFilms(searchUrl);
            }
        },
        [fetchFilms, searchUrl]
    );

    return (
        <header>
            <MainHeader>
                <HeaderText>{t('netflixRoulette')}</HeaderText>
                <SearchHeader>{t('header.searchText')}</SearchHeader>
                <SearchInput onInput={handleSearchInput} onKeyPress={handleEnter} />
                <ButtonsWrapper>
                    <SearchByButtons>
                        <SearchByText>{t('header.searchBy')}</SearchByText>
                        <Button
                            size={'small'}
                            onClick={handleSearchBy('title')}
                            active={searchByTitle}
                            margin
                            disabled={searchByTitle}
                        >
                            {t('header.title')}
                        </Button>
                        <Button
                            size={'medium'}
                            onClick={handleSearchBy('genre')}
                            active={searchByGenre}
                            disabled={searchByGenre}
                        >
                            {t('header.genre')}
                        </Button>
                    </SearchByButtons>
                    <Link href={`/search?search=${searchValue}&searchBy=${searchBy}`} as={`/search/${searchValue}`}>
                        <Button size={'big'} active disabled={searchValue === ''}>
                            {t('search')}
                        </Button>
                    </Link>
                </ButtonsWrapper>
            </MainHeader>
            <SubHeader>
                <Text>
                    {filmsCount} {t('header.found')}
                </Text>
                <Sorting>
                    <Text>{t('header.sortBy')}</Text>
                    <Text
                        onClick={handleSortBy('releaseDate')}
                        cursor={sortedByRating ? 'pointer' : ''}
                        color={sortedByReleaseDate ? '#d15c6e' : ''}
                        pointerEvents={sortedByReleaseDate ? 'none' : ''}
                    >
                        {t('header.releaseDate')}
                    </Text>
                    <Text
                        onClick={handleSortBy('rating')}
                        cursor={sortedByReleaseDate ? 'pointer' : ''}
                        color={sortedByRating ? '#d15c6e' : ''}
                        pointerEvents={sortedByRating ? 'none' : ''}
                    >
                        {t('header.rating')}
                    </Text>
                </Sorting>
            </SubHeader>
        </header>
    );
};

Header.propTypes = {
    filmsCount: PropTypes.number,
    sortBy: PropTypes.string,
    searchBy: PropTypes.string,
    toggleSorting: PropTypes.func,
    toggleSearchBy: PropTypes.func,
    fetchFilms: PropTypes.func
};

const mapStateToProps = state => getHeaderState(state);

const mapDispatchToProps = dispatch => ({
    toggleSorting: sortBy => dispatch(toggleSorting(sortBy)),
    toggleSearchBy: searchBy => dispatch(toggleSearchBy(searchBy)),
    fetchFilms: url => dispatch(getFilms(url))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Header);
