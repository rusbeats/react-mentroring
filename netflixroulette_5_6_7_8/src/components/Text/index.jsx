import React from 'react';
import PropTypes from 'prop-types';
import { StyledText } from './Text.styles';

const Text = ({ color, cursor, pointerEvents, children, onClick }) => {
    return (
        <StyledText onClick={onClick} color={color} cursor={cursor} pointerEvents={pointerEvents}>
            {children}
        </StyledText>
    );
};

StyledText.displayName = 'Text';

Text.propTypes = {
    color: PropTypes.string,
    cursor: PropTypes.string,
    pointerEvents: PropTypes.string,
    children: PropTypes.node,
    onClick: PropTypes.func
};

export default Text;
