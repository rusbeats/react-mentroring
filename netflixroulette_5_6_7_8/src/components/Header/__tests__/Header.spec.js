import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../index';
import { SearchInput } from '../Header.styles';
import Button from '../../Button';
import Text from '../../Text';

const PROPS = {
    sortBy: 'releaseDate',
    toggleSearchBy: jest.fn(),
    toggleSorting: jest.fn(),
    fetchFilms: jest.fn(),
    filmsCount: 0
};

describe('Header', () => {
    it('should render correctly (snapshot)', () => {
        const component = shallow(<Header />);
        expect(component).toMatchSnapshot();
    });

    it('should have buttons', () => {
        const component = shallow(<Header />);
        expect(component.find(Button)).toHaveLength(3);
    });

    it('should have texts', () => {
        const component = shallow(<Header />);
        expect(component.find(Text)).toHaveLength(4);
    });

    it('should have correct text props for sortBy releaseDate', () => {
        const component = shallow(<Header sortBy={PROPS.sortBy} />);
        const componentProps = component
            .find(Text)
            .at(2)
            .props();
        expect(componentProps).toEqual(
            expect.objectContaining({
                color: '#d15c6e',
                cursor: '',
                pointerEvents: 'none',
                children: 'header.releaseDate'
            })
        );
    });

    it('should have correct text props for sortBy rating', () => {
        PROPS.sortBy = 'rating';
        const component = shallow(<Header sortBy={PROPS.sortBy} />);
        const componentProps = component
            .find(Text)
            .at(2)
            .props();
        expect(componentProps).toEqual(
            expect.objectContaining({ color: '', cursor: 'pointer', pointerEvents: '', children: 'header.releaseDate' })
        );
    });

    it('should call toggleSearchBy with title', () => {
        const component = shallow(<Header toggleSearchBy={PROPS.toggleSearchBy} />);
        component
            .find(Button)
            .at(0)
            .simulate('click');
        expect(PROPS.toggleSearchBy).toHaveBeenCalledWith('title');
    });

    it('should call toggleSearchBy with genre', () => {
        const component = shallow(<Header toggleSearchBy={PROPS.toggleSearchBy} />);
        component
            .find(Button)
            .at(1)
            .simulate('click');
        expect(PROPS.toggleSearchBy).toHaveBeenCalledWith('genre');
    });

    it('should call fetchFilms on press Enter key', () => {
        const component = shallow(<Header fetchFilms={PROPS.fetchFilms} />);
        component.find(SearchInput).simulate('keyPress', { key: 'Enter' });
        expect(PROPS.fetchFilms).toHaveBeenCalled();
    });

    it('should not execute on not enter key', () => {
        const component = shallow(<Header fetchFilms={PROPS.fetchFilms} />);
        component.find(SearchInput).simulate('keyPress', { key: 'Space' });
        expect(PROPS.fetchFilms).toHaveReturned();
    });

    it('should call toggleSorting with releaseDate', () => {
        const component = shallow(<Header toggleSorting={PROPS.toggleSorting} />);
        component
            .find(Text)
            .at(2)
            .simulate('click');
        expect(PROPS.toggleSorting).toHaveBeenCalledWith('releaseDate');
    });

    it('should call toggleSorting with rating', () => {
        const component = shallow(<Header toggleSorting={PROPS.toggleSorting} />);
        component
            .find(Text)
            .at(3)
            .simulate('click');
        expect(PROPS.toggleSorting).toHaveBeenCalledWith('rating');
    });
});
