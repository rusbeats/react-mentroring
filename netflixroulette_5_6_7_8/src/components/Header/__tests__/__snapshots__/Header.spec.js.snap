// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly (snapshot) 1`] = `
<header>
  <Headerstyles__MainHeader>
    <Headerstyles__HeaderText>
      netflixRoulette
    </Headerstyles__HeaderText>
    <Headerstyles__SearchHeader>
      header.searchText
    </Headerstyles__SearchHeader>
    <Headerstyles__SearchInput
      onInput={[Function]}
      onKeyPress={[Function]}
    />
    <Headerstyles__ButtonsWrapper>
      <Headerstyles__SearchByButtons>
        <Headerstyles__SearchByText>
          header.searchBy
        </Headerstyles__SearchByText>
        <Button
          active={false}
          disabled={false}
          margin={true}
          onClick={[Function]}
          size="small"
        >
          header.title
        </Button>
        <Button
          active={false}
          disabled={false}
          onClick={[Function]}
          size="medium"
        >
          header.genre
        </Button>
      </Headerstyles__SearchByButtons>
      <Link
        as="/search/"
        href="/search?search=&searchBy=undefined"
      >
        <Button
          active={true}
          disabled={true}
          size="big"
        >
          search
        </Button>
      </Link>
    </Headerstyles__ButtonsWrapper>
  </Headerstyles__MainHeader>
  <Headerstyles__SubHeader>
    <Text>
       
      header.found
    </Text>
    <Headerstyles__Sorting>
      <Text>
        header.sortBy
      </Text>
      <Text
        color=""
        cursor=""
        onClick={[Function]}
        pointerEvents=""
      >
        header.releaseDate
      </Text>
      <Text
        color=""
        cursor=""
        onClick={[Function]}
        pointerEvents=""
      >
        header.rating
      </Text>
    </Headerstyles__Sorting>
  </Headerstyles__SubHeader>
</header>
`;
